app/api/queue.v2/infrastructure/specs/ApplicationRedisClient.spec.ts

```diff
@@ -0,0 +1,65 @@
+import EventEmitter from 'events';
+// eslint-disable-next-line camelcase
+import { ApplicationRedisClient, TESTING_clean } from '../ApplicationRedisClient';
+
+const clientMock = new EventEmitter();
+// @ts-ignore
+clientMock.quit = () => true;
+
+const createClientMock = jest.fn().mockReturnValue(clientMock);
+
+beforeEach(() => {
+  TESTING_clean();
+  clientMock.removeAllListeners();
+  createClientMock.mockClear();
+});
+
+it('should only call createClient once if many requests request an instance concurrently', done => {
+  Promise.all([
+    ApplicationRedisClient.getInstance(createClientMock),
+    ApplicationRedisClient.getInstance(createClientMock),
+    ApplicationRedisClient.getInstance(createClientMock),
+    ApplicationRedisClient.getInstance(createClientMock),
+    ApplicationRedisClient.getInstance(createClientMock),
+  ])
+    .then(([r1, r2, r3, r4, r5]) => {
+      expect(createClientMock.mock.calls.length).toBe(1);
+      expect(r1).toBe(r2);
+      expect(r1).toBe(r3);
+      expect(r1).toBe(r4);
+      expect(r1).toBe(r5);
+      expect(r1).toBe(clientMock);
+      done();
+    })
+    .catch(done);
+
+  clientMock.emit('ready');
+});
+
+it('should throw an error if the creation throws', done => {
+  const promise = ApplicationRedisClient.getInstance(createClientMock);
+  promise
```

Comment:
Even if we emitted the ready event, this is not flaky because we are not it's not a rogue promise, it's properly then-ed and catch-ed.
But yes, in any case, that would continue until the event is emitted. That means that if the code is broken, it may be scenarios in which we see it failing by timeout.